File: public/app/js/services/log.js
angular.module( 'portailApp' )
    .service( 'log',
              [ '$http', '$state', 'APP_PATH', 'currentUser',
                function( $http, $state, APP_PATH, currentUser ) {
                    this.add = function( app, url, params ) {
                        currentUser.get( false )
                            .then( function( user ) {
                                $http.post( APP_PATH + '/api/log',
                                            { app: app,
                                              uid: user.uid,
                                              uai: user.profil_actif.etablissement_code_uai,
                                              user_type: user.profil_actif.profil_id,
                                              timestamp: Date.now(),
                                              url: _(url).isNull() ? APP_PATH + $state.current.url: url,
                                              params: _(params).isNull() ? _($state.params).map( function( value, key ) { return key + '=' + value; } ).join( '&' ) : params } );
                            } );
                    };
                }
              ] );

File: routes/api/log.rb
# -*- coding: utf-8 -*-

module Portail
  module Routes
    module Api
      module Log
        def self.registered( app )
          app.get "#{APP_PATH}/api/log/?" do
            params[:askeruid] = user[:uid]

            AnnuaireWrapper::Log.query( params ).to_json
          end

          app.get "#{APP_PATH}/api/log/stats/?" do
            data = {}

            if params.key?( 'from' )
              data['from'] = params['from']
              data['until'] = params['until']
            end

            data['uid'] = user[:uid]

            stats = AnnuaireWrapper::Log.stats( data )

            uai_erasme = '0699990Z'
            stats['uai'].delete( uai_erasme )
            stats['general']['uai'].reject! { |item| item['uai'] == uai_erasme }

            stats.to_json
          end

          #
          # Tell Annuaire to log this
          #
          app.post "#{APP_PATH}/api/log/?" do
            # param :uid, String, required: true
            # param :uai, String, required: true
            # param :timestamp, BigNum, required: true
            # param :action, String, required: true
            # param :url, String, required: true
            # param :comment, String

            log_entry = JSON.parse( request.body.read )
            log_entry['ip'] = request.env[ 'HTTP_X_FORWARDED_FOR' ]
            AnnuaireWrapper::Log.add( log_entry )
          end
        end
      end
    end
  end
end
